### PR TITLE
Mail and mbox adapter (fixes #89)

### LINKED ISSUE #89: Idea: mailbox & maildir adapter
Hello

I was thinking, would it make sense to add adapters for parsing emails? They are almost textual, except that:

* The content might be encoded (quoted-printable, base64, …)
* The whole mbox file can also be optionally gzipped
* There can be various attachments (that could be pdf or ebooks...)

### PR DESCRIPTION
**Pull request by FliegendeWurst**
The new `MboxAdapter` provides the mail content to other extractors (this resolves #89).

### CODE REVIEW

**Comment by phiresky** [comment]
Hey. I'm a bit late here but if you would rebase this I would be happy to merge it. Maybe not too easy because adapters are async now..

**Comment by FliegendeWurst** [comment]
Hi, thanks for the response. I'm actually working on an `mbox` adapter too, which I will include in this PR once it's ready.

**Comment by phiresky** [comment]
uh good look with that. i tried writing a mbox parser in Rust a few years ago but apparently the mbox format has different incompatible versions that differ in how stuff is escaped and I couldn't even find docs about how the mbox format that Firefox uses works

**Comment by FliegendeWurst** [comment]
> uh good luck with that. i tried writing an mbox parser in Rust a few years ago but apparently the mbox format has different incompatible versions that differ in how the "from " line is escaped and I couldn't even find docs about how the mbox format that Firefox uses works

Oh just splitting on "\nFrom " seems to work for mbox files saved by Thunderbird. I think most mbox variantions forbid "\nFrom " in mail content.

File: src/adapters/email.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            }};
+        }
+        if self.buf_parts.is_none() {
+            let mut buf = Vec::new();
+            self.inp.inp.read_to_end(&mut buf).unwrap();
+            let queue_rc = Rc::new(RefCell::new(None));
+            self.buf_parts = Some((buf, queue_rc.clone()));
+            if let Ok(mail) = mailparse::parse_mail(&self.buf_parts.as_ref().unwrap().0) {
+                let mut queue = Vec::new();
+                let x = process_mail!(queue, mail);
```
[the remark this answers is not in the document]
Comment (FliegendeWurst) [reply src/adapters/email.rs]: This was a macro to recurse into mail subparts. In the new extractor, I just use a queue.

File: src/adapters/mbox.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            line_prefix,
+            archive_recursion_depth,
+            config,
+            postprocess,
+            ..
+        } = ai;
+
+        let mut content = Vec::new();
+        let s = stream! {
+            inp.read_to_end(&mut content).await?;
```
Comment (phiresky) [inline src/adapters/mbox.rs:70]: reading the whole file into RAM is not good because it may be many gigabytes in size (e.g. firefox mailbox files?)

**Pull request merged**